src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	ctx = clues.Add(ctx, "drive_id", uc.driveID)
+
+	// Lazy refresh
+	if uc.needsRefresh() {
+		err := uc.refreshCache(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	url, err := uc.readCache(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	return url, nil
+}
+
+// needsRefresh returns true if the cache is empty or if refresh interval has
+// elapsed
+func (uc *urlCache) needsRefresh() bool {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	return len(uc.urlMap) == 0 || time.Since(uc.lastRefreshTime) > uc.refreshInterval
+}
+
+// refreshCache refreshes the URL cache by performing a delta query.
+func (uc *urlCache) refreshCache(
+	ctx context.Context,
+) error {
+	// Acquire mutex to prevent multiple threads from refreshing the
+	// cache at the same time
+	uc.refreshMutex.Lock()
+	defer uc.refreshMutex.Unlock()
+
+	// If the cache was refreshed by another thread while we were waiting
+	// to acquire mutex, return
+	if !uc.needsRefresh() {
+		return nil
+	}
+
+	// Hold cache lock in write mode for the entire duration of the refresh.
+	// This is to prevent other threads from reading the cache while it is
+	// being updated page by page
+	uc.cacheLock.Lock()
+	defer uc.cacheLock.Unlock()
+
+	// Issue a delta query to graph
+	logger.Ctx(ctx).Debug("refreshing url cache")
+
+	err := uc.deltaQuery(ctx)
+	if err != nil {
+		return err
+	}
+
+	logger.Ctx(ctx).Debug("url cache refresh complete")
+	// Update last refresh time
+	uc.lastRefreshTime = time.Now()
+
+	return nil
+}
+
+// deltaQuery will perform a delta query on the drive. updateCache will be
+// called for each page of items returned by the delta query and the cache will
+// be updated with the results
+func (uc *urlCache) deltaQuery(
+	ctx context.Context,
+) error {
+	logger.Ctx(ctx).Debug("starting delta query")
+
+	_, _, _, err := collectItems(
+		ctx,
+		uc.itemPagerFunc(uc.svc, uc.driveID, ""),
+		uc.driveID,
+		"",
+		uc.updateCache,
+		map[string]string{},
+		"",
+		uc.errors,
+	)
+	if err != nil {
+		return clues.Wrap(err, "delta query failed").WithClues(ctx)
+	}
+
+	uc.deltaQueryCount++
+
+	logger.Ctx(ctx).Debug("finished delta query")
+
+	return nil
+}
+
+// readCache returns the download URL for the specified item
+func (uc *urlCache) readCache(
+	ctx context.Context,
+	itemID string,
+) (*itemProperties, error) {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	ctx = clues.Add(ctx, "item_id", itemID)
+
+	itemProps, ok := uc.urlMap[itemID]
+	if !ok {
+		return nil, clues.New("item not found in cache").WithClues(ctx)
+	}
+
+	return &itemProps, nil
+}
+
+// updateCache consumes a slice of drive items and updates the url cache.
+// It assumes that cacheLock is held by caller in write mode
+func (uc *urlCache) updateCache(
+	ctx context.Context,
+	_, _ string,
+	items []models.DriveItemable,
+	_ map[string]string,
+	_ map[string]string,
+	_ map[string]struct{},
+	_ map[string]map[string]string,
+	_ bool,
+	errs *fault.Bus,
```

Comment:
non-test funcs should only swallow parameters in very rare cases.  All of these ignored inputs produce a smell to me.  The caller won't easily know the difference between required and non-required values, and poor compliance with a currently-swallowed value can become a footgun if a maintainer updates only this func.  Is there something keeping us from removing these unused parameters?

Reply:
Fair point. Context here is that I am reusing `collectItems` func which which takes in a `collector` func which has above signature.

I had a cleaner approach in an [older commit](https://github.com/alcionai/corso/pull/3459/commits/549ae620129c9ec726b0a92cce016b99c0fb5430#diff-307fc0010b8fb64c7aa9cb85a4ac959aaadd612611595075b8a056ec93f8fa87L229)

But that was duplicating `collectItems` call. I looked into refactoring `collectItems` to separate delta query(`itempager`) code from `collector` but I didn't have any fresh ideas there. 

Let me switch back to the older commit, and file a tech debt item for `collectItems` refactoring.

Reply:
Honestly - I don't think we should duplicate collectItems (let's keep the enumeration logic in one place). Maybe a refactor as a follow up where we move these optional parameters into a struct

Reply:
@vkamra sounds good. Let me keep this code as above then. I'll file a tech debt item for `collectItems` refactoring as you mentioned.